docs/codeql/codeql-language-guides/analyzing-data-flow-in-swift.rst

```diff
@@ -0,0 +1,289 @@
+
+.. code-block:: ql
+
+   from DataFlow::Node source, DataFlow::Node sink
+   where MyDataFlow::flow(source, sink)
+   select source, "Dataflow to $@.", sink, sink.toString()
+
+Using global taint tracking
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Global taint tracking is to global data flow what local taint tracking is to local data flow.
+That is, global taint tracking extends global data flow with additional non-value-preserving steps.
+The global taint tracking library uses the same configuration module as the global data flow library but taint flow analysis is performed with ``TaintTracking::Global``:
+
+.. code-block:: ql
+
+   module MyTaintFlow = TaintTracking::Global<MyDataFlowConfiguration>;
+
+   from DataFlow::Node source, DataFlow::Node sink
+   where MyTaintFlow::flow(source, sink)
+   select source, "Taint flow to $@.", sink, sink.toString()
+
+Predefined sources
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The data flow library module ``codeql.swift.dataflow.FlowSources`` contains a number of predefined sources, providing a good starting point for defining data flow and taint flow based security queries.
+
+-  The class ``RemoteFlowSource`` represents data flow from remote network inputs and from other applications.
+-  The class ``LocalFlowSource`` represents data flow from local user input.
+-  The class ``FlowSource`` includes both of the above.
+
+Examples of global data flow
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following global taint-tracking query finds places where a string literal is used in a function call argument named "password".
+  - Since this is a taint-tracking query, the ``TaintTracking::Global`` module is used.
+  - The ``isSource`` predicate defines sources as any ``StringLiteralExpr``.
+  - The ``isSink`` predicate defines sinks as arguments to a ``CallExpr`` called "password".
+  - The sources and sinks may need tuning to a particular use, for example if passwords are represented by a type other than ``String`` or passed in arguments of a different name than "password".
+
+.. code-block:: ql
+
+   import swift
+   import codeql.swift.dataflow.DataFlow
+   import codeql.swift.dataflow.TaintTracking
+
+   module ConstantPasswordConfig implements DataFlow::ConfigSig {
+     predicate isSource(DataFlow::Node node) { node.asExpr() instanceof StringLiteralExpr }
+
+     predicate isSink(DataFlow::Node node) {
+       // any argument called `password`
+       exists(CallExpr call | call.getArgumentWithLabel("password").getExpr() = node.asExpr())
+     }
+
+   module ConstantPasswordFlow = TaintTracking::Global<ConstantPasswordConfig>;
+
+   from DataFlow::Node sourceNode, DataFlow::Node sinkNode
+   where ConstantPasswordFlow::flow(sourceNode, sinkNode)
+   select sinkNode, "The value '" + sourceNode.toString() + "' is used as a constant password."
```

Comment:
This would give the user clickable links to the source. Would that be better?

```suggestion
   select sinkNode, "The value $@ is used as a constant password.", sourceNode, sourceNode.toString()
```